core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -104,7 +98,45 @@ object RequestChannel extends Logging {
       }
     }
 
-    trace(s"Processor $processor received request: ${requestDesc(true)}")
+    def releaseBuffer(): Unit = {
+      if (buffer != null) {
+        memoryPool.release(buffer)
+        buffer = null
+      }
+    }
+
+    override def toString = s"Request(processor=$processor, " +
+      s"connectionId=${context.connectionId}, " +
+      s"session=$session, " +
+      s"listenerName=${context.listenerName}, " +
+      s"securityProtocol=${context.securityProtocol}, " +
+      s"buffer=$buffer)"
+  }
+
+  /** Summary of Request without the parsed request **/
+  class RequestSummary(
+    val processor: Int,
+    val metrics: Metrics,
+    val context: RequestContext,
+    val session: Session,
+    startTimeNanos: Long,
+    val description: String,
+    requestBodySizeInBytes: Int,
+    isFromFollower: Boolean
+  ) {
+    // These need to be volatile because the readers are in the network thread and the writers are in the request
+    // handler threads or the purgatory threads
+    @volatile var requestDequeueTimeNanos = -1L
+    @volatile var apiLocalCompleteTimeNanos = -1L
+    @volatile var responseCompleteTimeNanos = -1L
+    @volatile var responseDequeueTimeNanos = -1L
+    @volatile var apiRemoteCompleteTimeNanos = -1L
+    @volatile var messageConversionsTimeNanos = 0L
+    @volatile var temporaryMemoryBytes = 0L
+    @volatile var recordNetworkThreadTimeCallback: Option[Long => Unit] = None
```

Comment:
I think we can add these vars to a separate class and then we can include an instance of this class in `Request` _and_ `RequestSummary`. That way, we don't have to force the creation of `RequestSummary` to access these fields.

Reply:
Just re-read your comment and you were asking if this was worth doing. I think it is.

Reply:
Cool.  I created a `RequestTimers` type as agreed.  I also made the `RequestSummary` creation as late as possible.  It's now only created in the instantiation of the `Response`'s in `KafkaApis`.  Everything looked good until I ran the test suite and found an interesting regression revealed by the `kafka.server.RequestQuotaTest.testUnauthorizedThrottle` test.

A behaviour of this test is that it cycles through different request types while a request quota is enforced.  When request quota is enforced and the request is exceeding it, `ClientQuotaManager` will throttle the creation of the response.  The problem is that some request types never trigger the lazy evaluation to parse the `AbstractRequest` and the buffer gets released before its had a chance to evaluate, causing NPE's for the `RequestSummary` bits that depend on it (`description`, `sizeInBytes`, etc.) when eventually the response is created (https://github.com/seglo/kafka/blob/to-request-summary-5859/core/src/main/scala/kafka/server/KafkaRequestHandler.scala#L72).

An example of a request that can cause this regression is one with the ApiKey `SASL_AUTHENTICATE`.  Most requests handled in `KafkaApis` call the `Request.body` method which parses and returns the concrete request type, but I assume this (and probably others) don't have or need an implementation.

This wasn't a problem before because the `AbstractRequest` was eagerly evaluated and available in the `RequestChanne.Request` throttle callback to create a `Response`.  

Some possible solutions:

1. Always force evaluation of `bodyAndSize` in the `Request.releaseBuffer` method when the buffer is not null.
2. Identify which requests don't trigger the lazy evaluation and make them do so even though it's not required at that time.  IMO this is more brittle than 1., because it would need to be considered for each new request type added in the future.

Sorry for the novel.  It was an interesting troubleshooting experience and I wanted to reinforce the knowledge by documenting it here!  I haven't implemented any solutions yet because I wanted to give you a chance to weigh in.